internal/collector/csr.go

```diff
@@ -0,0 +1,151 @@
+	descCSRLabelsHelp          = "Kubernetes labels converted to Prometheus labels."
+	descCSRLabelsDefaultLabels = []string{"csr"}
+
+	csrMetricFamilies = []metric.FamilyGenerator{
+		{
+			Name: descCSRLabelsName,
+			Type: metric.MetricTypeGauge,
+			Help: descCSRLabelsHelp,
+			GenerateFunc: wrapCSRFunc(func(j *certv1beta1.CertificateSigningRequest) metric.Family {
+				labelKeys, labelValues := kubeLabelsToPrometheusLabels(j.Labels)
+				return metric.Family{
+					Metrics: []*metric.Metric{
+						{
+							LabelKeys:   labelKeys,
+							LabelValues: labelValues,
+							Value:       1,
+						},
+					},
+				}
+			}),
+		},
+		{
+			Name: "kube_csr_created",
+			Type: metric.MetricTypeGauge,
+			Help: "Unix creation timestamp",
+			GenerateFunc: wrapCSRFunc(func(csr *certv1beta1.CertificateSigningRequest) metric.Family {
+				ms := []*metric.Metric{}
+				if !csr.CreationTimestamp.IsZero() {
+					ms = append(ms, &metric.Metric{
+						LabelKeys:   []string{},
+						LabelValues: []string{},
+						Value:       float64(csr.CreationTimestamp.Unix()),
+					})
+				}
+
+				return metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+		{
+			Name: "kube_csr_condition",
+			Type: metric.MetricTypeGauge,
+			Help: "Condition of the csr",
+			GenerateFunc: wrapCSRFunc(func(csr *certv1beta1.CertificateSigningRequest) metric.Family {
+				return metric.Family{
+					Metrics: addCSRConditionMetrics(csr.Status),
+				}
+			}),
+		},
+	}
+)
+
+func wrapCSRFunc(f func(*certv1beta1.CertificateSigningRequest) metric.Family) func(interface{}) metric.Family {
+	return func(obj interface{}) metric.Family {
+		csr := obj.(*certv1beta1.CertificateSigningRequest)
+
+		metricFamily := f(csr)
+
+		for _, m := range metricFamily.Metrics {
+			m.LabelKeys = append(descCSRLabelsDefaultLabels, m.LabelKeys...)
+			m.LabelValues = append([]string{csr.Name}, m.LabelValues...)
+		}
+
+		return metricFamily
+	}
+}
+
+func createCSRListWatch(kubeClient clientset.Interface, ns string) cache.ListWatch {
+	return cache.ListWatch{
+		ListFunc: func(opts metav1.ListOptions) (runtime.Object, error) {
+			return kubeClient.CertificatesV1beta1().CertificateSigningRequests().List(opts)
+		},
+		WatchFunc: func(opts metav1.ListOptions) (watch.Interface, error) {
+			return kubeClient.CertificatesV1beta1().CertificateSigningRequests().Watch(opts)
+		},
+	}
+}
+
+// addCSRConditionMetrics generates one metric for each possible csr condition status
+func addCSRConditionMetrics(cs certv1beta1.CertificateSigningRequestStatus) []*metric.Metric {
+	approved := false
+	denied := false
+	if len(cs.Conditions) > 0 {
+		for _, s := range cs.Conditions {
+			if s.Type == certv1beta1.CertificateApproved {
+				approved = true
+			}
+			if s.Type == certv1beta1.CertificateDenied {
+				denied = true
+			}
+		}
+	}
+	//"approved" if a certificate request has the
+	// "Approved" condition and no "Denied" conditions; false otherwise.
+	approved = approved && !denied
```

Comment:
interpretation of data should be done at query time, not in kube-state-metrics, we should just expose each state with a 0 or 1 depending on whether it is in that state

Reply:
A little confused, what do you mean by "query time" ? do you mean I can just remove this line?  when the csr  have "approved" and "denied" condition at the same time,  should I expose the metrics both 1 like this:
kube_csr_condition{csr="certificate-test",csr_condition="denied"} 1
kube_csr_condition{csr="certificate-test",csr_condition="approved"} 1

Reply:
@brancz if we don't interpret the data, there is the possibility for both approved and denied being true.

This upstream section seems to be the source of truth: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/certificates/certificate_controller_utils.go#L23

Reply:
Correct, but here we want to expose the raw data. Essentially the `approved && !denied` is something that is up to a user to do in PromQL.

Reply:
This make sense, and it's how I did originally